### PR TITLE
feat: adds last active timestamp

### LINKED ISSUE #17153: [CAL-4540] track "last active" per user in user table (takes last booking or last usage of app.cal.com)
for both analytics but also as a column in the "Member List"

image

<sub>[CAL-4540](https://linear.app/calcom/issue/CAL-4540/track-last-login-per-user-in-user-table)</sub>

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

- Fixes https://github.com/calcom/cal.com/issues/17153
- Fixes CAL-4540

This pull request introduces a feature to track the last active time of users and updates various components to display this information. The most important changes include updating the user model, modifying repositories to handle the new field, and updating UI components to display the last active time.

### Database and Model Changes:
* Added `lastActiveAt` column to the `users` table in the database schema. (`packages/prisma/migrations/20241023041541_add_user_last_active_at/migration.sql`)
* Updated the `User` model to include the `lastActiveAt` field. (`packages/prisma/schema.prisma`)

### Repository Updates:
* Created `UsersRepository` with a method to update `lastActiveAt` for a user. (`packages/features/users/users.repository.ts`)
* Defined `IUsersRepository` interface for the new repository method. (`packages/features/users/users.repository.interface.ts`)

### Backend Logic:
* Updated `getServerSession` to set `lastActiveAt` when a user session is retrieved. (`packages/features/auth/lib/getServerSession.ts`)
* Modified `handleNewBooking` to update `lastActiveAt` for users associated with new bookings. (`packages/features/bookings/lib/handleNewBooking.ts`)

### UI and API Changes:
* Added `lastActiveAt` column to `UserListTable` and `MemberList` components. (`packages/features/users/components/UserTable/UserListTable.tsx`) [[1]](diffhunk://#diff-53ea1951f4dfbf097b8d923cddbdc947753f3ae9ebf5b18b6188311b9ca6b7ebR325-R329) (`packages/features/ee/teams/components/MemberList.tsx`) [[2]](diffhunk://#diff-0fc75c34f34292d3cabc33fcd1ee97ef37b880ba7ec49d50a9f51ec849a34c93R380-R384)
* Updated API handlers to include `lastActiveAt` in user data responses. (`packages/trpc/server/routers/viewer/organizations/listMembers.handler.ts`) [[1]](diffhunk://#diff-f3d594c1e43f45a9446335c61110116e516f8b4a1eb35599b325564498cc5645R109) [[2]](diffhunk://#diff-f3d594c1e43f45a9446335c61110116e516f8b4a1eb35599b325564498cc5645R168-R174) (`packages/trpc/server/routers/viewer/teams/listMembers.handler.ts`) [[3]](diffhunk://#diff-4888d0d376a49acb3dc97b1ea2f265848730659ef1a70e21a6f9ed7e3e2c88f8R28) [[4]](diffhunk://#diff-4888d0d376a49acb3dc97b1ea2f265848730659ef1a70e21a6f9ed7e3e2c88f8R96-R102)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [x] I have updated the developer docs in /docs if this PR makes changes that would require a [documentation change](https://cal.com/docs). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my changes generate no new warnings

### CODE REVIEW

File: packages/features/auth/lib/getServerSession.ts [hunk 1]
```diff
@@ -54,10 +54,11 @@ export async function getServerSession(options: {
     return cachedSession;
   }
 
-  const userFromDb = await prisma.user.findUnique({
+  const userFromDb = await prisma.user.update({
```
Comment (zomars) [inline packages/features/auth/lib/getServerSession.ts:57]: We update the last active timestamp and retrieve the user in one go

Comment (zomars) [reply packages/features/auth/lib/getServerSession.ts:57]: This covers both login and each session retrieval

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 2]
```diff
@@ -950,6 +951,11 @@ async function handler(
       originalRescheduledBooking,
     });
 
+    if (booking?.userId) {
+      const usersRepository = new UsersRepository();
+      await usersRepository.updateLastActiveAt(booking.userId);
+    }
```
Comment (zomars) [inline packages/features/bookings/lib/handleNewBooking.ts:957]: If the booking is associated with a user we update the active timestamp

File: packages/features/users/users.repository.interface.ts
Comment (zomars) [inline packages/features/users/users.repository.interface.ts:1]: Basic repository to move away from direct prisma usage

File: packages/trpc/server/routers/viewer/organizations/listMembers.handler.ts [hunk 3]
```diff
@@ -164,6 +165,13 @@ export const listMembersHandler = async ({ ctx, input }: GetOptions) => {
         accepted: membership.accepted,
         disableImpersonation: user.disableImpersonation,
         completedOnboarding: user.completedOnboarding,
+        lastActiveAt: membership.user.lastActiveAt
+          ? new Intl.DateTimeFormat(ctx.user.locale, {
+              timeZone: ctx.user.timeZone,
+            })
+              .format(membership.user.lastActiveAt)
+              .toLowerCase()
+          : null,
```
Comment (zomars) [inline packages/trpc/server/routers/viewer/organizations/listMembers.handler.ts:174]: Dates come pre-formated from server:

image

**Review by zomars — COMMENTED** [review]
Self review done

File: packages/features/ee/teams/components/MemberList.tsx [hunk 4]
```diff
@@ -377,6 +377,11 @@ export default function MemberList(props: Props) {
           return filterValue.includes(rows.getValue(id));
         },
       },
+      {
+        id: "lastActiveAt",
+        header: "Last Active",
```
Comment (PeerRich) [inline packages/features/ee/teams/components/MemberList.tsx:382]: eventually we have to translate these but i see all headers are hard coded english right now

Comment (zomars) [reply packages/features/ee/teams/components/MemberList.tsx:382]: I think a separate PR so translate all headers is the way to go

**Review by PeerRich — COMMENTED** [review]
reduced priority for next week ~ medium

**Review by anikdhabal — APPROVED** [review]

**Pull request merged**